Dockerfile

```diff
@@ -2,82 +2,63 @@
 # Zulip development environment image and use
 # tools/build-release-tarball to generate a production release tarball
 # from the provided Git ref.
-FROM ubuntu:xenial-20171114
-LABEL maintainer="Alexander Trost <galexrt@googlemail.com>"
+FROM ubuntu:18.04 as base
 
-# You can specify these in docker-compose.yml or with
-#   docker build --build-args "ZULIP_GIT_REF=git_branch_name" .
-ARG ZULIP_GIT_URL=https://github.com/zulip/zulip.git
-ARG ZULIP_GIT_REF=2.0.4
-ARG CUSTOM_CA_CERTIFICATES=
+# Set up working locales and upgrade the base image
+ENV LANG="C.UTF-8"
 
-SHELL ["/bin/bash", "-xuo", "pipefail", "-c"]
+ARG UBUNTU_MIRROR
 
-# First, we setup working locales
-RUN echo 'APT::Install-Recommends 0;' >> /etc/apt/apt.conf.d/01norecommends && \
-    echo 'APT::Install-Suggests 0;' >> /etc/apt/apt.conf.d/01norecommends && \
+RUN { [ ! "$UBUNTU_MIRROR" ] || sed -i "s|http://\(\w*\.\)*archive\.ubuntu\.com/ubuntu/\? |$UBUNTU_MIRROR |" /etc/apt/sources.list; } && \
     apt-get -q update && \
-    apt-get -q install locales && \
-    locale-gen en_US.UTF-8
+    apt-get -q dist-upgrade -y && \
+    DEBIAN_FRONTEND=noninteractive \
+        apt-get -q install --no-install-recommends -y ca-certificates git locales lsb-release python3 sudo tzdata
 
-ENV LANG="en_US.UTF-8" \
-    LANGUAGE="en_US:en" \
-    LC_ALL="en_US.UTF-8"
 
-# Next, we upgrade the base image and add a zulip user
-RUN apt-get -q update && \
-    apt-get -q dist-upgrade -y && \
-    apt-get -q install -y git sudo ca-certificates apt-transport-https python3 crudini && \
-    useradd -d /home/zulip -m zulip && \
+FROM base as build
+
+# Add a zulip user
+RUN useradd -d /home/zulip -m zulip && \
     echo 'zulip ALL=(ALL:ALL) NOPASSWD:ALL' >> /etc/sudoers
 
+USER zulip
+WORKDIR /home/zulip
+
+# You can specify these in docker-compose.yml or with
+#   docker build --build-args "ZULIP_GIT_REF=git_branch_name" .
+ARG ZULIP_GIT_URL=https://github.com/zulip/zulip.git
+ARG ZULIP_GIT_REF=2.0.4
```

Comment:
Can we put these a little higher in the file?  Since they're the main user-facing interface, it seems like it'd be better to have them above the `useradd` stuff.

Reply:
The position of `ARG` in a Dockerfile is semantically significant. Providing an `ARG` invalidates all the layers below it, so you want it to be as far down as possible.

We can document in comments near the top, if you want.

Reply:
I guess we already document specifying these via `docker-compose.yml`, so it's probably fine.